ocaml/xapi/xapi_vm_helpers.ml

```diff
@@ -58,6 +58,30 @@ let ensure_domain_type_is_specified ~__context ~self =
     derive_domain_type ~hVM_boot_policy:hbp |> fun value ->
     Db.VM.set_domain_type ~__context ~self ~value
 
+let ensure_boot_params_consistent ~__context ~self =
+  (* XSI-804 if we are missing an HVM boot order, then qemu won't
+   * be able to start the VM successfully *)
+  if Db.VM.get_domain_type ~__context ~self = `hvm then
+    let boot_params = Db.VM.get_HVM_boot_params ~__context ~self in
+    let msg =
+      match List.assoc_opt "order" boot_params with
+      | None ->
+          Some "is missing HVM_boot_params.order"
+      | Some "" ->
+          Some "has empty HVM_boot_params.order"
+      | Some _ ->
+          None
+    in
+    match msg with
+    | None ->
+        ()
+    | Some msg ->
+        raise
+          Api_errors.(
+            Server_error
+              ( invalid_value
+              , [Printf.sprintf "VM=%s %s" (Ref.short_string_of self) msg] ))
```

Comment:
This error is defined as `error Api_errors.invalid_value ["field"; "value"]` in datamodel_errors.ml, so you need something like `["HVM_boot_params:order", value_from_db]`.

Reply:
By the way, we are not at all consistent with these errors in our code base :(